### PR TITLE
Create HoverEvent from ItemStack Entity

### PR DESCRIPTION
**Pull request by yannicklamprecht**

### CODE REVIEW

**Comment by LuminousLoreBloom** [comment]
~~:thumbsup:~~

~~I'd just wait for this to be pulled first so you can use this constructor: https://github.com/SpigotMC/BungeeCord/pull/2889/files#diff-4935e96c7fd911acac330566a446db74R46~~

~~Then you can remove ItemTag, and just convert the Bukkit ItemStack to an NMS ItemStack, save it to a NBTTagCompound and parse it through the constructor as a toString().~~

**Review by LuminousLoreBloom — DISMISSED** [review]

File: Spigot-API-Patches/0218-Create-HoverEvent-from-ItemStack-Entity.patch
Comment (yannicklamprecht) [inline Spigot-API-Patches/0218-Create-HoverEvent-from-ItemStack-Entity.patch]: Should we wait until this is non snapshot and available upstream? Could probably come at that time when 1.16.2 releases?

**Comment by MiniDigger** [comment]
FYI: bungee has been released https://github.com/SpigotMC/BungeeCord/commit/a99f62f693825b429c56353d46a6494bd1d4b2b8

**Comment by yannicklamprecht** [comment]
Done.

File: Spigot-API-Patches/0218-Create-HoverEvent-from-ItemStack-Entity.patch
Comment (A248) [inline Spigot-API-Patches/0218-Create-HoverEvent-from-ItemStack-Entity.patch]: Shouldn't this accept a `BaseComponent[]` as do other component APIs?

Comment (BillyGalbreath) [reply Spigot-API-Patches/0218-Create-HoverEvent-from-ItemStack-Entity.patch]: Not everything takes an array. Some things require a singular component. You can wrap your array of components in a singular one like so:
```new TextComponent(components)```

**Comment by yannicklamprecht** [comment]
Any updates on these? It's just a little change and it's ready for merge.

File: Spigot-Server-Patches/0544-Create-HoverEvent-from-ItemStack-Entity.patch
Comment (l1m0n3) [inline Spigot-Server-Patches/0544-Create-HoverEvent-from-ItemStack-Entity.patch]:
```suggestion
+        ItemTag itemTag = ItemTag.ofNbt(CraftItemStack.asNMSCopy(itemStack).getTag().toString());
```

Comment (KennyTV) [reply Spigot-Server-Patches/0544-Create-HoverEvent-from-ItemStack-Entity.patch:36]: No

Comment (yannicklamprecht) [reply Spigot-Server-Patches/0544-Create-HoverEvent-from-ItemStack-Entity.patch]: ~~No that would not work.
That's what is called currently. By just getting the tag you would drop the type and the count.~~


Yeah, you're right.

Comment (l1m0n3) [reply Spigot-Server-Patches/0544-Create-HoverEvent-from-ItemStack-Entity.patch]: I just noticed `getTag()` returns null when the item doesn't have any additional data (e.g. enchantments, lores)
so, you have to check if the tag is null or not before `toString()`

Comment (yannicklamprecht) [reply Spigot-Server-Patches/0544-Create-HoverEvent-from-ItemStack-Entity.patch]: Thanks. Fixed by using `getOrCreateTag()` instead.

File: Spigot-Server-Patches/0551-Create-HoverEvent-from-ItemStack-Entity.patch
Comment (Proximyst) [inline Spigot-Server-Patches/0551-Create-HoverEvent-from-ItemStack-Entity.patch]:
```suggestion
+        return hoverContentOf(entity, StringUtils.isBlank(customName) ? null : new TextComponent(customName));
```

File: Spigot-API-Patches/0219-Create-HoverEvent-from-ItemStack-Entity.patch
Comment (Proximyst) [inline Spigot-API-Patches/0219-Create-HoverEvent-from-ItemStack-Entity.patch]:
```suggestion
+     * @param customName a custom name that should be displayed, if not passed entity name will be displayed
+     * @return the {@link net.md_5.bungee.api.chat.hover.content.Content} of that {@link Entity}
```

File: Spigot-API-Patches/0219-Create-HoverEvent-from-ItemStack-Entity.patch
Comment (Proximyst) [inline Spigot-API-Patches/0219-Create-HoverEvent-from-ItemStack-Entity.patch]:
```suggestion
+     * @param customName a custom name that should be displayed, if not passed entity name will be displayed
+     * @return the {@link net.md_5.bungee.api.chat.hover.content.Content} of that {@link Entity}
```

**Review by Proximyst — DISMISSED** [review]
Just minor nitpicks. LGTM!

**Comment by yannicklamprecht** [comment]
Will do a squash and rebase tomorrow. 6pm CEST August 25th

**Comment by LuminousLoreBloom** [comment]
Just thought I'd like to add:
There's a PR on the Spigot hub by one of the maintainers there that implements **only** the Item Hover API that was published a couple of days ago; it might be a good idea to wait for md's judgement on pulling that in; to pull this completely if he does not approve, or otherwise; modify this PR to **only** retain the Entity Hover API which emulates the method structure of the other PR?
Otherwise it might cause a bit of a railway split between the two.

**Comment by yannicklamprecht** [comment]
Rebased onto master and squashed with Co-Author.

File: Spigot-API-Patches/0221-Create-HoverEvent-from-ItemStack-Entity.patch
Comment (Proximyst) [inline Spigot-API-Patches/0221-Create-HoverEvent-from-ItemStack-Entity.patch]: Should this not be `BaseComponent[]`?

File: Spigot-API-Patches/0221-Create-HoverEvent-from-ItemStack-Entity.patch
Comment (Proximyst) [inline Spigot-API-Patches/0221-Create-HoverEvent-from-ItemStack-Entity.patch]: Should this not be `BaseComponent[]`?

File: Spigot-API-Patches/0221-Create-HoverEvent-from-ItemStack-Entity.patch
Comment (yannicklamprecht) [reply Spigot-API-Patches/0221-Create-HoverEvent-from-ItemStack-Entity.patch]: We have only one TextComponent that is one BaseComponent.
https://github.com/SpigotMC/BungeeCord/blob/master/chat/src/main/java/net/md_5/bungee/api/chat/hover/content/Entity.java#L36

File: Spigot-API-Patches/0221-Create-HoverEvent-from-ItemStack-Entity.patch
Comment (yannicklamprecht) [reply Spigot-API-Patches/0221-Create-HoverEvent-from-ItemStack-Entity.patch]: We have only one TextComponent that is one BaseComponent.
https://github.com/SpigotMC/BungeeCord/blob/master/chat/src/main/java/net/md_5/bungee/api/chat/hover/content/Entity.java#L36

File: Spigot-API-Patches/0221-Create-HoverEvent-from-ItemStack-Entity.patch
Comment (Proximyst) [reply Spigot-API-Patches/0221-Create-HoverEvent-from-ItemStack-Entity.patch]: Perhaps it'd be nice to add a new method to accept `BaseComponent[]` that simply adds them as extras? This is because `ComponentBuilder#create` returns `BaseComponent[]` (as do many other methods to construct them).

Comment (yannicklamprecht) [reply Spigot-API-Patches/0221-Create-HoverEvent-from-ItemStack-Entity.patch]: Yea, will do that after uni paper stuff.

Comment (A248) [reply Spigot-API-Patches/0221-Create-HoverEvent-from-ItemStack-Entity.patch]: funny, I asked the same thing

Comment (yannicklamprecht) [reply Spigot-API-Patches/0221-Create-HoverEvent-from-ItemStack-Entity.patch]: Yeah, but Proxy provided an example where it is reasonable.

**Review by Proximyst — DISMISSED** [review]

File: Spigot-API-Patches/0221-Create-HoverEvent-from-ItemStack-Entity.patch
Comment (aikar) [inline Spigot-API-Patches/0221-Create-HoverEvent-from-ItemStack-Entity.patch]: These methods belong on ItemFactory, no reason to pollute the root API with these.

**Review by aikar — DISMISSED** [review]
one minor nitpick, if either you or the dev team can move these, good w/ me.

going to pre approve with note they need to be moved first before dev team merges.
@PaperMC/development-team

File: Spigot-Server-Patches/0565-Create-HoverEvent-from-ItemStack-Entity.patch
Comment (LuminousLoreBloom) [inline Spigot-Server-Patches/0565-Create-HoverEvent-from-ItemStack-Entity.patch]: Maybe also another overload method would be nice:
hoverContentOf(Entity) which refers to the entity's already set custom name or just pass null if hasCustomName==false
https://papermc.io/javadocs/paper/1.16/org/bukkit/Nameable.html#getCustomName--

Comment (MiniDigger) [reply Spigot-Server-Patches/0565-Create-HoverEvent-from-ItemStack-Entity.patch]: added

File: Spigot-API-Patches/0221-Create-HoverEvent-from-ItemStack-Entity.patch
Comment (MiniDigger) [reply Spigot-API-Patches/0221-Create-HoverEvent-from-ItemStack-Entity.patch]: done

**Comment by MiniDigger** [comment]
@limbo-app ready

**Review by limbo-app — APPROVED** [review]
Pull request automatically approved for merging - a core team member has previously approved this pull request.

**Comment by MiniDigger** [comment]
thanks!

**Pull request merged**
